typhos/benchmark/utils.py

```diff
@@ -0,0 +1,159 @@
+from contextlib import contextmanager
+from inspect import isclass, isfunction
+from multiprocessing import Process
+
+from ophyd.signal import EpicsSignalBase
+
+logger = logging.getLogger(__name__)
+
+
+_optional_err = ('Optional dependency caproto missing from python '
+                 'environment. Cannot test server.')
+
+try:
+    from caproto.server import pvproperty, PVGroup, run
+    has_caproto = True
+except ImportError:
+    has_caproto = False
+    logger.debug(_optional_err)
+
+
+def run_caproto_ioc(device_class, prefix):
+    """
+    Runs a dummy caproto IOC.
+
+    Includes all the PVs that device_class will have if instantiated with
+    prefix.
+
+    Assumes only basic :class:`ophyd.Component` instances in the class
+    definition.
+    """
+    if not has_caproto:
+        raise ImportError(_optional_err)
+
+    pvprops = {}
+    for suffix in yield_all_suffixes(device_class):
+        pvprops[suffix] = pvproperty()
+
+    DynIOC = type('DynIOC', (PVGroup,), pvprops)
+    ioc = DynIOC(prefix)
+
+    run(ioc.pvdb, module_name='caproto.asyncio.server',
+        interfaces=['0.0.0.0'])
+
+
+def yield_all_suffixes(device_class):
+    """
+    Iterates through all full pvname suffixes defined by device_class.
+
+    Assumes only basic :class:`ophyd.Component` instances in the class
+    definition.
+    """
+    for walk in device_class.walk_components():
+        if issubclass(walk.item.cls, EpicsSignalBase):
+            suffix = get_suffix(walk)
+            yield suffix
+
+
+def get_suffix(walk):
+    """
+    Returns the full pvname suffix from a ComponentWalk instance
+
+    This means everything after the top-level device's prefix.
+    Assumes that walk is an :class:`ophyd.signal.EpicsSignalBase`
+    subclass and that it was defined using only
+    :class:`ophyd.Component` in the device ancestors tree.
+    """
+    suffix = ''
+    for cls, attr in zip(walk.ancestors, walk.dotted_name.split('.')):
+        suffix += getattr(cls, attr).suffix
+    return suffix
+
+
+@contextmanager
+def caproto_context(device_class, prefix):
+    """
+    Yields a caproto process with all elements of the input device.
+
+    The caproto IOC will be run in a background process, making it suitable for
+    testing devices in the main process.
+    """
+    if not has_caproto:
+        raise ImportError(_optional_err)
+
+    proc = Process(target=run_caproto_ioc, args=(device_class, prefix))
+    proc.start()
+    yield
+    if proc.is_alive():
+        os.kill(proc.pid, signal.SIGKILL)
+
+
+def random_prefix():
+    """Returns a random prefix to avoid test cross-talk."""
+    return str(uuid.uuid4())[:8] + ':'
+
+
+def is_native(obj, module):
+    """Returns True if obj was defined in module."""
+    return module.__name__ in obj.__module__
+
+
+def get_native_functions(module):
+    """Returns all functions and methods defined in module."""
+    return get_native_methods(module, module)
+
+
+def get_native_methods(cls, module):
```

Comment:
Basically, a loop if two classes share mutual references to each other? (Or a similar less direct loop?) Seems common enough to be an issue.

Reply:
Yeah, some cyclical reference - at any depth. Try the original method on `caproto` and that should trigger it.

Reply:
And then I assume the `native_methods` passing is just for a speedup so we create fewer lists... there's also an issue I see now where we might recurse through way too many objects. I'll also edit this to call `is_native` sooner so we don't end up in some other module and spider through someone else's package (obviously finding nothing)

Reply:
That sounds about right. It was easier to paste my quick fixes that allowed it to work for profiling of caproto rather than make a poor attempt at explaining it 🤷‍♂️